### PR TITLE
New protocol: gemini

### PR DESCRIPTION
**Pull request by KAction**
This patch implements gemini protocol as described in specification[^1],
with following pieces missing:

 * Redirects. 3x status codes are not handled any specially from any
   other status code that has no body, -L option has no effect on gemini
   URLs.

Other than that, things work:

	$ make
	$ ./src/curl -k -D- gemini://tilde.pink/~kaction/dist/
	20 text/gemini
	# Directory listing

	=> /~kaction ..
	=> flake-dhall/ flake-dhall/                                           Nov 22 2020

[^1]: https://gemini.circumlunar.space/docs/specification.html

### CODE REVIEW

File: lib/gemini.c [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+   * functions and avoid busy loops.
+   */
+
+  return CURLE_OK;
+}
+
+static CURLcode gemini_doing_finish(struct connectdata *, bool *);
+static CURLcode gemini_doing(struct connectdata *conn, bool *done)
+{
+  CURLcode err;
```
Comment (bagder) [inline lib/gemini.c:94]: If you'd make the `CURLcode' variables be named 'result' you would follow standard curl code patterns closer. Not a requirement, just an idea.

File: lib/gemini.c [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+     * For many servers first read returns exacly header, because it is
+     * natural thing to do on server side, although we can't rely on it.
+     * But this is reason why it does not worth to optimize search by
+     * keeping track of old {amount} value and searching only in bytes
+     * just read.
+     */
+    gemini->block.lf = memchr(gemini->block.memory, '\n',
+                              gemini->block.amount);
+    if(!gemini->block.lf && gemini->block.done)
+      return CURLE_WEIRD_SERVER_REPLY;
```
Comment (bagder) [inline lib/gemini.c:162]: maybe add a failf() call and explain the reason for the failure?

File: lib/gemini.c [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+#include "curl_setup.h"
+
+#if !defined CURL_DISABLE_GEMINI && defined USE_SSL
+
+#include <ctype.h>
+#include <string.h>
+#include "gemini.h"
+#include "urldata.h"
+#include "vtls/vtls.h"
+#include <curl/curl.h>
```
Comment (bagder) [inline lib/gemini.c:32]: curl/curl.h is already included from within urldata.h

File: lib/gemini.c [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+#include "transfer.h"
+#include "sendf.h"
+#include "connect.h"
+#include "multiif.h"
+#include "progress.h"
+#include "gopher.h"
```
Comment (bagder) [inline lib/gemini.c:38]: surely gopher.h isn't needed here?

**Comment by bagder** [comment]
Test 1119 fails because `docs/libcurl/symbols-in-versions` needs an update.

Test 1165 fails because: Used in code, not documented in CURL-DISABLE.md: CURL_DISABLE_GEMINI

File: lib/gemini.c [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+  }
+
+  /* stage2: read block big enough to contain header */
+  if(!gemini->block.done) {
+    char *into;
+
+    into = gemini->block.memory + gemini->block.amount;
+    more = GEMINI_RESPONSE_BUFSIZE - gemini->block.amount;
+
+    SOCKET_READABLE(sockfd, -1);
```
[the remark this answers is not in the document]
Comment (KAction) [reply lib/gemini.c:129]: As I read it, SOCKET_READABLE is macro which expands to call of Curl_socket_check, which "Wait for read or write events on a set of file descriptors". Last argument is timeout, negative means wait indefinitely.

So this line does exactly what I want, but in blocking way, which I am not happy with either.

https://github.com/curl/curl/blob/c7b02c5d680021f21aebdf41cb336a742fdad664/lib/select.c#L224
https://github.com/curl/curl/blob/c7b02c5d680021f21aebdf41cb336a742fdad664/lib/select.h#L84

File: lib/gemini.c (hunk 4 above)
Comment (KAction) [reply lib/gemini.c]: Definitely not. Cleaned up headers.

File: lib/gemini.c (hunk 2 above)
Comment (KAction) [reply lib/gemini.c:165]: Added explanation for every instance of CURLE_WEIRD_SERVER_REPLY.

File: lib/gemini.c (hunk 1 above)
Comment (KAction) [reply lib/gemini.c]: Renamed.

File: lib/gemini.c [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+     * invokactions of "doing" function, so it is desirable to get
+     * things done in as little calls to "doing" function, as possible,
+     * but without busy looping on socket that is not yet ready.
+     *
+     * For many servers first read returns exacly header, because it is
+     * natural thing to do on server side, although we can't rely on it.
+     * But this is reason why it does not worth to optimize search by
+     * keeping track of old {amount} value and searching only in bytes
+     * just read.
+     */
```
[the remark this answers is not in the document]
Comment (KAction) [reply lib/gemini.c:153]: Applied your wording fix.

**Comment by KAction** [comment]
Looks like I broke a lot of things, but I don't understand what exactly went wrong. For example,

https://curlbuild.uxnr.de/builders/curl_winssl_cross_x86_dbg/builds/13151

There is seems to be no logs for failing stage.

**Comment by KAction** [comment]
I figured out how to implement redirects. @bagder Would you prefer another curl flag, or re-using CURLOPT_FOLLOWLOCATION is fine? If you want new flag, would your prefer new flag for command line tool or to have -L set two flags at same time?

Also, I discovered that `curl gemini://.... | less ` sometimes fails with `socket not ready` error; converting pull request to draft.

File: lib/gemini.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (bagder) [reply lib/gemini.c]: Sorry, I confused the arguments to the function. The -1 is the timeout value, which should be 0 to prevent it from waiting.

**Comment by bagder** [comment]
> re-using CURLOPT_FOLLOWLOCATION is fine?

Yes, I think this is the option for redirect-following no matter which protocol!

**Comment by mback2k** [comment]
> https://curlbuild.uxnr.de/builders/curl_winssl_cross_x86_dbg/builds/13151

See the testcurl step before: https://curlbuild.uxnr.de/builders/curl_winssl_cross_x86_dbg/builds/13151/steps/testcurl/logs/stdio

**Comment by bagder** [comment]
Is this still being worked on? If not, I don't want to keep it open anymore...

**Pull request closed without merging**